compact_str/src/lib.rs

```diff
@@ -737,3 +732,51 @@ impl Add<CompactStr> for String {
 }
 
 crate::asserts::assert_size_eq!(CompactStr, String);
+
+/// A trait for converting a value to a `CompactStr`.
+///
+/// This trait is automatically implemented for any type which implements the
+/// [`Display`] trait. As such, `ToCompactStr` shouldn't be implemented directly:
+/// [`Display`] should be implemented instead, and you get the `ToCompactStr`
+/// implementation for free.
+///
+/// [`Display`]: fmt::Display
+pub trait ToCompactStr {
+    /// Converts the given value to a `CompactStr`.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use compact_str::ToCompactStr;
+    ///
+    /// let i = 5;
+    /// let five = "5".to_compact_str();
+    ///
+    /// assert_eq!(five, i.to_string());
+    /// ```
+    fn to_compact_str(&self) -> CompactStr;
+}
+
+/// # Panics
+///
+/// In this implementation, the `to_compact_str` method panics
+/// if the `Display` implementation returns an error.
+/// This indicates an incorrect `Display` implementation
+/// since `std::fmt::Write for String` never returns an error itself and
+/// the implementation of `ToCompactStr::to_compact_str` only panic if the `Display`
+/// implementation is incorrect..
+impl<T: fmt::Display> ToCompactStr for T {
+    #[inline]
+    fn to_compact_str(&self) -> CompactStr {
+        #[cfg(feature = "to-compact-str-specialisation")]
+        if let Some(compact_str) = to_compact_str_specialisation::to_compact_str_specialised(self) {
+            return compact_str;
+        }
+
+        CompactStr {
+            repr: Repr::from_fmt(self),
```

Comment:
Can we implement `fmt::Write` for `Repr`, and then here do something like:
```
let num_bytes = count(&self);
let mut repr = Repr::with_capacity(num_bytes);
write!(&mut repr, "{}", self).expect(...);
```

Reply:
Done!

I've taken things even further, by using `CompactStr::with_capacity` then use `fmt::Write` implementation for `CompactStr` to format `self` into `compact_str.`